src/features/class-full-page-cache-404.php

```diff
@@ -0,0 +1,337 @@
+ *
+ * (c) Alley <info@alley.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ *
+ * @package Alley\WP\WP_404_Caching
+ */
+
+declare( strict_types=1 );
+
+namespace Alley\WP\WP_404_Caching\Features;
+
+use Alley\WP\Types\Feature;
+use WP_Query;
+use function defined;
+use function function_exists;
+
+/**
+ * Full Page Cache for 404s.
+ */
+final class Full_Page_Cache_404 implements Feature {
+
+	/**
+	 * Cache group.
+	 *
+	 * @var string
+	 */
+	public const CACHE_GROUP = 'wp_404_caching';
+
+	/**
+	 * Cache key.
+	 *
+	 * @var string
+	 */
+	public const CACHE_KEY = '404_cache';
+
+	/**
+	 * Cache key for stale cache.
+	 *
+	 * @var string
+	 */
+	public const STALE_CACHE_KEY = '404_cache_stale';
+
+	/**
+	 * Stale cache time.
+	 *
+	 * @var int
+	 */
+	public const STALE_CACHE_TIME = DAY_IN_SECONDS;
+
+	/**
+	 * Cache time.
+	 *
+	 * @var int
+	 */
+	public const DEFAULT_CACHE_TIME = HOUR_IN_SECONDS;
+
+	/**
+	 * Stale cache time.
+	 *
+	 * @var int
+	 */
+	public const DEFAULT_STALE_CACHE_TIME = DAY_IN_SECONDS;
+
+	/**
+	 * Guaranteed 404 URI.
+	 * Used for populating the cache.
+	 *
+	 * @var string
+	 */
+	public const TEMPLATE_GENERATOR_URI = '/wp-404-caching/404-template-generator/?generate=1&uri=1';
+
+	/**
+	 * Get cache time.
+	 *
+	 * @return int
+	 */
+	public static function get_cache_time(): int {
+		return apply_filters( 'wp_404_caching_cache_time', self::DEFAULT_CACHE_TIME );
+	}
+
+	/**
+	 * Get stale cache time.
+	 *
+	 * @return int
+	 */
+	public static function get_stale_cache_time(): int {
+		return apply_filters( 'wp_404_caching_stale_cache_time', self::DEFAULT_STALE_CACHE_TIME );
+	}
+
+	/**
+	 * Boot the feature.
+	 */
+	public function boot(): void {
+
+		/**
+		 * Only boot feature if external object cache is being used.
+		 *
+		 * We don't want to store the cached 404 page in the database.
+		 *
+		 * And only boot feature if the site is using SSL.
+		 */
+		if ( ! wp_using_ext_object_cache() || ! is_ssl() ) {
+			return;
+		}
+
+		// Return 404 page cache on template_redirect.
+		add_action( 'template_redirect', [ self::class, 'action__template_redirect' ], 1 );
+
+		// For the Guaranteed 404 page, hook in on WP to start output buffering, to capture the HTML.
+		add_action( 'wp', [ self::class, 'action__wp' ] );
+
+		// Replenish the cache every hour.
+		if ( ! wp_next_scheduled( 'wp_404_cache' ) ) {
+			wp_schedule_event( time(), 'hourly', 'wp_404_cache' );
+		}
+
+		// Callback for Cron Event.
+		add_action( 'wp_404_cache', [ self::class, 'trigger_404_page_cache' ] );
+		add_action( 'wp_404_cache_single', [ self::class, 'trigger_404_page_cache' ] );
+	}
+
+	/**
+	 * Get 404 Page Cache and return early if found.
+	 */
+	public static function action__template_redirect(): void {
+
+		// Don't cache if user is logged in.
+		if ( is_user_logged_in() ) {
+			return;
+		}
+
+		// Don't cache if not a 404.
```

Comment:
In my mind this would be the first condition. I know we don't want to cache for either of these, but not being a 404 seems like it would be the quickest escape hatch.

Reply:
Good call!